Tag Builder: init

Horki opened the pull request:
@museun hi, this is an initial version of Tag Builder, I created in a special namespace as a helper

TODO:
- should values be unique by key?
- if not, can we have multiple key-value values?
- what sort of sanitization should we have when building tags? (remove empty space, allow ':', ',' in key-value?)

Horki committed:
Tag Builder: init

museun commented:
* keys should be unique. I imagine 'last' key wins (the normal insertion rules for a hashmap)
    * note, IRCv3 allows for duplicate keys but Twitch does not (yet? they don't send duplicate keys)
* there are some quoting rules about tags, I'll have to find them. the Twitch docs are notably lacking in many areas.

There should be a `merge` function so you can take a pre-parsed message and modify its existed tags with a new set of tags.

This is what I came up with, but there are various edge cases I haven't bothered to handle. I've been exceedingly busy lately:
https://github.com/museun/still_shaken/blob/master/crates/still_shaken/src/bot/test.rs#L264-L318


I think the one provided in this crate should operate directly on `Tags/TagsIndices` though.

museun commented on src/irc/tag_indices.rs:
```diff
@@ -95,3 +103,84 @@ impl IntoOwned<'static> for TagIndices {
+        {
+            // TODO: should we be able to append new data after build?
+            let r = tag.with("first", "second").build();
+            assert_eq!(r, "#first:second,first:second,first:second");
```
The format for tags is `@key1=val1;key2=val2; `

museun committed:
reimplement tags builder

this introduces a 'test' module which'll be a place for user-facing
useful utilities.

the tags builder can both incrementally build tags and merge existing
tags.

this requires escaping and unescaping strings. so two helper functions
are provided.

in the near future get() will unescape by default -- but currently
get_unescaped() provides that role.

museun commented:
I ended up rewriting it using some of your design.

I had to do some reworking to support escaping and unescaping. 

I apologize for all of my rewrites of your contributions. This crate is structured weirdly (with a lot of hidden 'magic') that makes it non-trivial (or obvious, to be honest) on how to do things.

Your contributions are motivating me, though. This 'feature' would've stayed unimplemented forever because I would just do it locally for whatever I use this crate for.

museun commented:
As for your questions: I decided to implement it according to the IRCv3 spec: 
* keys are unique 
    * but it is not an error to replace a key
* values can be empty
* it escapes the keys and values when it encodes the tags
* this won't be exposed off existing types, but through a new 'test' module.
    * the builder would only be used for users to create test messages for their own use.
    * the crate, internally will build up messages outside of tags (via the 'commands' module).

museun commented on src/test/tags_builder.rs:
```diff
@@ -0,0 +1,259 @@
+    /// This returns a tuple that can be used with [`Tags::from_data_indices`](crate::irc::Tags#from_data_indices)
+    ///
+    /// If any empty keys were found, or no keys at all then an error will be returned,
+    pub fn build(self) -> Result<(MaybeOwned<'static>, TagIndices), BuilderError> {
```
From a user perspective this return type is a bit .. lets say weird.

I don't if I should make it more user-friendly -- but Tags are build from a data reference and pre-computed indices.

I could maybe return a simple helper struct which could have a `fn tags(&self) -> Tags<'_>` on it.

museun replied:
I implemented this in 920448e

museun committed:
add a nicer return type for tags builder

museun commented:
I'm satisfied with the design. Merging this

Horki commented:
@museun that's a lot of work in a short period :+1:
